packages/@lwc/template-compiler/src/config.ts

```diff
@@ -26,20 +34,62 @@ export interface Config {
     preserveHtmlComments?: boolean;
 }
 
-export type NormalizedConfig = Required<Config>;
+export type NormalizedConfig = Required<Omit<Config, 'customRendererConfig'>> &
+    Partial<Pick<Config, 'customRendererConfig'>>;
 
 const AVAILABLE_OPTION_NAMES = new Set([
+    'customRendererConfig',
     'experimentalComputedMemberExpression',
     'experimentalDynamicDirective',
     'preserveHtmlComments',
 ]);
 
+function normalizeCustomRendererConfig(config: CustomRendererConfig): CustomRendererConfig {
+    const tagNames: string[] = [];
+    const normalizedConfig: CustomRendererConfig = {
+        elements: config.elements.map((e) => {
+            const tagName = e.tagName.toLowerCase();
+            // Custom element cannot be allowed to have a custom renderer hook
+            // The renderer is cascaded down from the owner(custom element) to all its child nodes who
+            // do not have a renderer specified.
+            invariant(!isCustomElementTag(tagName), TemplateErrors.CUSTOM_ELEMENT_TAG_DISALLOWED, [
+                e.tagName,
+            ]);
+
+            tagNames.push(tagName);
+            return {
+                tagName,
+                namespace: e.namespace?.toLowerCase(),
+                attributes: e.attributes?.map((a) => a.toLowerCase()),
+            };
+        }),
+        directives: config.directives.map((d) => d.toLowerCase()),
```

Comment:
Why all this normalization is needed? Should we ask developers to pass correct names from the get-go?

Reply:
TBH, I don't have a strong opinion either way. This was going the extra step to cover for any developer mistakes with tag name casing. For example, the browser treats tag names in a case-insensitive way. `document.createElement('div')` `document.createElement('DIV')` `document.createElement('Div')`, all produce the same result.